portal-impl/src/com/liferay/portlet/internal/PortletRequestImpl.java

```diff
@@ -677,27 +703,39 @@ else if (themeDisplay.isLifecycleAction()) {
 			facesPortlet = true;
 		}
 
+		int portletSpecMajorVersion = portletApp.getSpecMajorVersion();
+
+		Set<String> privateRenderParameterNames = new LinkedHashSet<>();
+
 		if (portletFocus) {
-			Map<String, String[]> renderParameters = null;
+			Map<String, String[]> privateRenderParameters = null;
 
 			Map<String, String[]> parameters = request.getParameterMap();
 
 			for (Map.Entry<String, String[]> entry : parameters.entrySet()) {
 				RequestParameter requestParameter = new RequestParameter(
-					entry.getKey(), entry.getValue(), portletNamespace);
+					entry.getKey(), entry.getValue(), portletNamespace,
+					portletSpecMajorVersion);
 
 				if (requestParameter.isNameInvalid(_strutsPortlet)) {
 					continue;
 				}
 
-				if (themeDisplay.isLifecycleRender()) {
-					if (renderParameters == null) {
-						renderParameters = new HashMap<>();
+				if (getLifecycle().equals(PortletRequest.HEADER_PHASE) ||
```

Comment:
@shuyangzhou The reason why I switched from `themeDisplay.isLifecycleRender()` to `getLifecycle().equals(PortletRequest.RENDER_PHASE)` is because of the new concern about the header phase.

During development, it became clear to me that it didn't make sense to add a `ThemeDisplay.isLifecycleHeader()` method because the underlying boolean values in `ThemeDisplay` are set by checking the value of the `p_p_lifecycle` URL parameter. As you know, 0=render, 1=action, 2=resource. But it didn't make sense to add a 3=header type of parameter value, because there is no such thing as a `HeaderURL` or a `MimeResponse.createHeaderURL()` method. Instead, `RenderURL` first invokes the `HEADER_PHASE`, then the `RENDER_PHASE`. There is no type of URL that will simply/only invoke the `HEADER_PHASE` without invoking the `RENDER_PHASE`.